File: 0001-0500/0310_findMinHeightTrees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n <= 2: return list(range(n))
        graph = collections.defaultdict(set)
        for a, b in edges:
            graph[a].add(b)
            graph[b].add(a)
        leaves = [key for key, val in graph.items() if len(val) == 1]
        seen = {i for i in leaves}
        queue = collections.deque([leaves])
        while len(graph) > 2:
            level = queue.popleft()
            newLevel = []
            for leaf in level:
                candidate = graph[leaf].pop()
                del graph[leaf]
                graph[candidate].remove(leaf)
                if len(graph[candidate]) == 1 and candidate not in seen:
                    seen.add(candidate)
                    newLevel.append(candidate)
            if newLevel:
                queue.append(newLevel)
        return queue[0]

# Alternative solution

class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n <= 2:
            return list(range(n))

        graph = {i: [] for i in range(n)}
        in_degree = {i: 0 for i in range(n)}

        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
            in_degree[a] += 1
            in_degree[b] += 1
        
        queue = deque([key for key, val in in_degree.items() if val == 1])

        MHT = []

        while queue:
            level_length = len(queue)
            
            for _ in range(level_length):
                node = queue.popleft()
                in_degree[node] -= 1

                for neighbor in graph[node]:
                    if in_degree[neighbor] == 0:
                        continue

                    in_degree[neighbor] -= 1

                    if in_degree[neighbor] == 1:
                        queue.append(neighbor)

            if queue:
                MHT = list(queue)
            
        return MHT

File: 0001-0500/0310_findMinHeightTrees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        ArrayList<Integer>[] graph = new ArrayList[n];
        int[] indegree = new int[n];
        Queue<Integer> queue = new LinkedList<>();
        int remaining = n;

        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] edge: edges) {
            int u = edge[0];
            int v = edge[1];
            graph[u].add(v);
            graph[v].add(u);
            indegree[u]++;
            indegree[v]++;
        }

        for (int i = 0; i < n; i++) {
            if (indegree[i] <= 1) {
                queue.add(i);
            }
        }
        
        while (remaining > 2) {
            int length = queue.size();

            for (int i = 0; i < length; i++) {
                int node = queue.remove();
                remaining--;

                for (int child: graph[node]) {
                    indegree[child]--;

                    if (indegree[child] == 1) {
                        queue.add(child);
                    }
                }
            }
        }

        return new ArrayList(queue);
    }
}
